### PR TITLE
Install knative-kafka in Kyma Runtime

### PR DESCRIPTION
**Pull request by anishj0shi**
w.r.t https://github.com/kyma-project/kyma/issues/7421
Dynamic provisioning of Azure Event- hubs from the kyma environment Broker per provisioned kyma-runtime. 

- [x]  Provision Azure Event Hubs using HAP (Hyperscaler Account Pool)
- [x] Inject Secret of an azure event hub namespace to the kyma-runtime as an installation override.
- [x] Inject ConfigMap for switching knative's default Channel to **KafkaChannel**

### Some Follow ups:
- Use a standard Naming Convention to create Azure Resource Group in Kyma Environment Broker https://github.com/kyma-incubator/compass/issues/967
- Create an Azure Event-hub instance in the same location as the provisioned kyma-runtime https://github.com/kyma-incubator/compass/issues/968

### CODE REVIEW

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	log.Infof("HAP lookup for credentials to provision cluster for global account ID %s on Hyperscaler %s", pp.ErsContext.GlobalAccountID, hypType)
+
+	// get hyperscaler credentials from HAP
+	credentials, err := p.accountProvider.GardenerCredentials(hypType, pp.ErsContext.GlobalAccountID)
+	if err != nil {
+		errorMessage := fmt.Sprintf("Unable to retrieve Gardener Credentials from HAP lookup: %v", err)
+		return p.retryOperation(operation, errorMessage, gardenerCredentialsRetryInterval, gardenerCredentialsMaxTime, log)
+	}
+	azureCfg, err := azure.GetConfigfromHAPCredentialsAndProvisioningParams(credentials, pp)
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: err not handled

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 2]
```diff
@@ -129,16 +132,20 @@ func main() {
...
 	runtimeStep := provisioning.NewCreateRuntimeStep(db.Operations(), db.Instances(), provisionerClient)
 	smOverrideStep := provisioning.NewServiceManagerOverridesStep(db.Operations(), cfg.ServiceManager)
 
 	logs := logrus.New()
 	stepManager := process.NewManager(db.Operations(), logs)
 	stepManager.InitStep(initialisation)
+
 	stepManager.AddStep(1, resolveCredentialsStep)
-	stepManager.AddStep(2, smOverrideStep)
+	stepManager.AddStep(2, provisionAzureEventHub)
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/cmd/broker/main.go]: is it also valid for GCP provider?

Comment (anishj0shi) [reply components/kyma-environment-broker/cmd/broker/main.go]: Hi PM, 
yes currently this is the default eventing system for managed offering. so it's applicable for all.

Comment (piotrmiskiewicz) [reply components/kyma-environment-broker/cmd/broker/main.go]: So if I'm spinning Kyma Runtime in GCP azure event hub also will be provisioned, right?

Comment (anishj0shi) [reply components/kyma-environment-broker/cmd/broker/main.go]: yup, till we come up with an alternative for GCP. or something generic for both ;)

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go (hunk 1 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: done

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func extractEndpoint(accessKeys *eventhub.AccessKeys) string {
+	endpoint := strings.Split(*accessKeys.PrimaryConnectionString, ";")[0]
+	endpoint = strings.TrimPrefix(endpoint, "Endpoint=sb://")
+	endpoint = strings.TrimSuffix(endpoint, "/")
+	return endpoint
+}
+
+func appendPort(endpoint string, port int64) string {
```
Comment (k15r) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: you should be good with `port int`

File: components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+	return &fakeHyperscalerProvider{
+		client: client,
+		err:    nil,
+	}
+}
+
+func NewFakeHyperscalerProviderError() azure.HyperscalerProvider {
+	return &fakeHyperscalerProvider{
+		client: nil,
+		err:    fmt.Errorf("ups ... "),
```
Comment (k15r) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go]: I assume you do not expect this to happen

File: components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+
+	// when
+	op, _, err = step.Run(op, fixLogger())
+	require.NoError(t, err)
+	provisionRuntimeInput, err := op.InputCreator.Create()
+	require.NoError(t, err)
+
+	// then
+	allOverridesFound := ensureOverrides(t, provisionRuntimeInput)
+	require.True(t, allOverridesFound[componentNameKnativeEventing], "overrides for %s were not found", componentNameKnativeEventing)
```
Comment (k15r) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go]: `assert.True` here would print both failures

File: components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+
+func ensureOperationIsNotRepeated(t *testing.T, err error, when time.Duration, op internal.ProvisioningOperation) {
+	t.Helper()
+	require.NotNil(t, err)
+}
+
+// ensureOverrides ensures that the overrides for
+// - the kafka channel controller
+// - and the default knative channel
+// is set
```
Comment (k15r) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go]:
```suggestion
// are set
```

File: components/kyma-environment-broker/internal/event-hub/azure/env.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	config.clientID = clientID
+	config.clientSecret = clientSecret
+	config.tenantID = tenantID
+	config.subscriptionID = subscriptionID
+	config.location = location
+	return config, nil
+}
+
+func GetConfigfromHAPCredentialsAndProvisioningParams(credentials hyperscaler.Credentials, parameters internal.ProvisioningParameters) (*Config, error) {
+	region := "westeurope"
```
[the remark this answers is not in the document]
Comment (anishj0shi) [reply components/kyma-environment-broker/internal/event-hub/azure/env.go]: there's a follow up: https://github.com/kyma-incubator/compass/issues/968

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 8]
```diff
@@ -129,16 +132,20 @@ func main() {
...
 	stepManager.AddStep(10, runtimeStep)
+	stepManager.AddStep(3, smOverrideStep)
```
Comment (jasiu001) [inline components/kyma-environment-broker/cmd/broker/main.go]: nit: please keep order :)

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
+package event_hub
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: all provisioning step should be inside `/internal/process/provisioning/` dir

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+	if err != nil {
+		// if the parameters are incorrect, there is no reason to retry the operation
+		// a new request has to be issued by the user
+		log.Errorf("Aborting after failing to get valid operation provisioning parameters: %v", err)
+		return p.operationManager.OperationFailed(operation, "invalid operation provisioning parameters")
+	}
+	log.Infof("HAP lookup for credentials to provision cluster for global account ID %s on Hyperscaler %s", pp.ErsContext.GlobalAccountID, hypType)
+
+	// get hyperscaler credentials from HAP
+	credentials, err := p.accountProvider.GardenerCredentials(hypType, pp.ErsContext.GlobalAccountID)
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: `GardenerCredentials` method underneath fetch secret or update it. Those operations could be temporary, so maybe `time.Duration` here should be returned? Also, it is good practice to not run steps forever (when you return `time.Duration`) so at the beginning of step `operation.UpdatedAt` field can be checked how long operation is in progress state.
Another question is, it is necessary step to make provisioning, if not step can return success with proper log error.

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+	if err != nil {
+		// retrying might solve the issue while communicating with azure, e.g. network problems etc
+		errorMessage := fmt.Sprintf("Failed to persist Azure Resource Group [%s] with error: %v", groupName, err)
+		return p.retryOperation(operation, errorMessage, time.Minute, time.Minute*30, log)
+	}
+	log.Printf("Persisted Azure Resource Group [%s]", groupName)
+
+	// create EventHubs Namespace
+	eventHubsNamespace := pp.Parameters.Name
+	eventHubNamespace, err := namespaceClient.CreateNamespace(p.context, azureCfg, groupName, eventHubsNamespace)
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: to be sure, if the method will be called second time with the same params, update operation will be run and no error will be return?

File: components/kyma-environment-broker/internal/event-hub/azure/namespace_test.go [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+func TestGetResourceGroup(t *testing.T) {
+
+	resourceGroup := "my-resourcegroup"
+	eventHubNamespaceID := fmt.Sprintf("/subscriptions/35d42578-34d1-486d-a689-012a8d514c19/resourceGroups/%s/providers/Microsoft.EventHub/namespaces/nachtmaar-test-2", resourceGroup)
+	nameSpaceFixture := eventhub.EHNamespace{
+		Response: autorest.Response{},
+		ID:       &eventHubNamespaceID,
+	}
+	parsedResourceGroup := GetResourceGroup(nameSpaceFixture)
+	if parsedResourceGroup != resourceGroup {
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/event-hub/azure/namespace_test.go]: you can just use `assert.Equal(t, resourceGroup, parsedResourceGroup)` from "github.com/stretchr/testify/assert" library

File: components/kyma-environment-broker/internal/process/operation.go [hunk 13]
```diff
@@ -19,17 +19,21 @@ func NewOperationManager(storage storage.Operations) *OperationManager {
 	return &OperationManager{storage: storage}
 }
 
+// OperationSucceeded marks the operation as failed and only repeats it if there is a storage error
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/process/operation.go]: "marks the operation as failed" ? :)

File: components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+	provisionRuntimeInput, err := op.InputCreator.Create()
+	require.NoError(t, err)
+
+	// then
+	allOverridesFound := ensureOverrides(t, provisionRuntimeInput)
+	require.True(t, allOverridesFound[componentNameKnativeEventing], "overrides for %s were not found", componentNameKnativeEventing)
+	require.True(t, allOverridesFound[componentNameKnativeEventingKafka], "overrides for %s were not found", componentNameKnativeEventingKafka)
+}
+
+func Test_StepProvisionParametersError(t *testing.T) {
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go]: please use table test if it is possible

File: components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go (hunk 4 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go]: I expect this to happen, because it triggers the edge case in the eventhub step.

It is used in the test [here](https://github.com/anishj0shi/compass/blob/00ea98a7139d8d0c954ee8899f43dd2ab1a22bb5/components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go#L222)

It is testing the following lines:

```go
	namespaceClient, err := p.hyperscalerProvider.GetClient(azureCfg)
	if err != nil { <- it produces this error
		// internal error, repeating doesn't solve the problem
		errorMessage := fmt.Sprintf("Failed to create Azure EventHubs client: %v", err)
		return p.operationManager.OperationFailed(operation, errorMessage)
	}
```

Comment (k15r) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go]: well then lets use some proper message here :D

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+	// get EventHubs Namespace secret
+	accessKeys, err := namespaceClient.GetEventhubAccessKeys(p.context, *resourceGroup.Name, *eventHubNamespace.Name, authorizationRuleName)
+	if err != nil {
+		// retrying might solve the issue while communicating with azure, e.g. network problems etc
+		errorMessage := fmt.Sprintf("Unable to retrieve access keys to azure event-hub namespace: %v", err)
+		return p.retryOperation(operation, errorMessage, time.Minute, time.Minute*30, log)
+	}
+	kafkaEndpoint := extractEndpoint(&accessKeys)
+	kafkaEndpoint = appendPort(kafkaEndpoint, kafkaPort)
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: Good point, but doesn't work I think.
I wanted to create an instance of the struct (provide host and port) and call .String(), but the struct doesn't have a port entry.

> type URL struct {
>     Scheme     string
>     Opaque     string    // encoded opaque data
>     User       *Userinfo // username and password information
>     Host       string    // host or host:port
>     Path       string    // path (relative paths may omit leading slash)
>     RawPath    string    // encoded path hint (see EscapedPath method); added in Go 1.5
>     ForceQuery bool      // append a query ('?') even if RawQuery is empty; added in Go 1.7
>     RawQuery   string    // encoded query values, without '?'
>     Fragment   string    // fragment for references, without '#'
> }

File: components/kyma-environment-broker/internal/event-hub/azure/iam.go [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+		if err != nil {
+			return nil, err
+		}
+		a = autorest.NewBearerAuthorizer(token)
+
+	default:
+		return a, fmt.Errorf("invalid grant type specified")
+	}
+
+	return a, err
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/azure/iam.go]: Removed the switch and the enum. Currently we only need this particular OAuth Grant.

File: components/kyma-environment-broker/internal/process/operation.go (hunk 13 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/process/operation.go]: Ah, copy paste. Thank you!

File: components/kyma-environment-broker/internal/event-hub/azure/namespace_test.go (hunk 12 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/azure/namespace_test.go]: Method is not required anymore. Thanks for the suggestion 👍

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go (hunk 11 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: Good catch!

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go (hunk 10 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: Good catch. I didn't see the last call in the implementation. You are right, it makes sense to retry.

Concerning UpdatedAt, we implemented `retryOperation` which makes use of this field. Moving this to a common package could help others.

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go (hunk 11 above)
Comment (anishj0shi) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: if the method is called for the second time with the same parameter, it will not return any error. it would just try to update it.  
ref: https://github.com/Azure/azure-sdk-for-go/blob/f1b3b9b7d4962c94c77d16ce75f9a8e981eb2488/services/eventhub/mgmt/2017-04-01/eventhub/namespaces.go#L132

File: components/kyma-environment-broker/internal/event-hub/azure/group.go [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+)
+
+// getGroupsClient gets a client for handling of Azure ResourceGroups
+func getGroupsClient(config *Config) (resources.GroupsClient, error) {
+	client := resources.NewGroupsClient(config.subscriptionID)
+
+	authorizer, err := GetResourceManagementAuthorizer(config)
+	if err != nil {
+		return resources.GroupsClient{}, fmt.Errorf("failed to initialize authorizer with error: %v", err)
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/azure/group.go]: I had the same idea, but CI is preventing us from doing so. 


See this pipeline: https://storage.googleapis.com/kyma-prow-logs/pr-logs/pull/kyma-incubator_compass/905/pre-compass-components-kyma-environment-broker/1239571009451855873/build-log.txt
```
internal/event-hub/azure/client.go:27:15: Errorf format %w has unknown verb w
```

Compass is not using `%w` yet in the code. As soon as they do so, we can change it :)

File: components/kyma-environment-broker/internal/event-hub/azure/group.go [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+	}
+	client.Authorizer = authorizer
+
+	if err = client.AddToUserAgent(config.userAgent); err != nil {
+		return resources.GroupsClient{}, fmt.Errorf("failed to add use agent [%s] with error: %v", config.userAgent, err)
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/azure/group.go]: Won't do. See comment above. But thanks for the suggestion 👍

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+	log logrus.FieldLogger) (internal.ProvisioningOperation, time.Duration, error) {
+
+	hypType := hyperscaler.Azure
+
+	// parse provisioning parameters
+	pp, err := operation.GetProvisioningParameters()
+	if err != nil {
+		// if the parameters are incorrect, there is no reason to retry the operation
+		// a new request has to be issued by the user
+		log.Errorf("Aborting after failing to get valid operation provisioning parameters: %v", err)
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: Good question!

The problem only appears if the provisioning parameters cannot be unmarshalled (e.g. invalid json). The request to the KymaEnvironmentBroker (KEB) is send via the UI later, so I hope the UI will display a reasonable error message. 

The step operation is async, so ideally the provisioning parameters (which are part of the HTTP PUT request), are validated before the async operation started and give the user a 4XX status code. But to be honest, I don't know if this is the case. 

Since the code `operation.GetProvisioningParameters()` is used heavily throughout KEB, I would leave it to the authors of KEB to implement fast user feedback :)

File: components/kyma-environment-broker/internal/event-hub/event_hub_step.go (hunk 9 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step.go]: Done. Furthermore, I moved the azure package to `/internal/hyperscaler`. Let me know if you have a different suggestion :)

File: components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go (hunk 4 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/event-hub/event_hub_step_test.go]: all ups are more detailed now :-D

File: components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+	// get EventHubs Namespace secret
+	accessKeys, err := namespaceClient.GetEventhubAccessKeys(p.context, *resourceGroup.Name, *eventHubNamespace.Name, authorizationRuleName)
+	if err != nil {
+		// retrying might solve the issue while communicating with azure, e.g. network problems etc
+		errorMessage := fmt.Sprintf("Unable to retrieve access keys to azure event-hub namespace: %v", err)
+		return p.retryOperation(operation, errorMessage, time.Minute, time.Minute*30, log)
+	}
+	kafkaEndpoint := extractEndpoint(accessKeys)
+	kafkaEndpoint = appendPort(kafkaEndpoint, kafkaPort)
+	kafkaPassword := *accessKeys.PrimaryConnectionString
```
Comment (mszostok) [inline components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go]: please check the pointer as such expression can lead to panic

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 21]
```diff
@@ -129,15 +132,19 @@ func main() {
...
 	smOverrideStep := provisioning.NewServiceManagerOverridesStep(db.Operations(), cfg.ServiceManager)
 
 	logs := logrus.New()
 	stepManager := process.NewManager(db.Operations(), logs)
 	stepManager.InitStep(initialisation)
+
 	stepManager.AddStep(1, resolveCredentialsStep)
-	stepManager.AddStep(2, smOverrideStep)
+	stepManager.AddStep(2, provisionAzureEventHub)
+	stepManager.AddStep(3, smOverrideStep)
```
Comment (mszostok) [inline components/kyma-environment-broker/cmd/broker/main.go]: Getting the SM overrides are not connected with azure hub, so they be done in parallel, am I right @jasiu001?

File: components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+
+	return operation, 0, nil
+}
+
+func (p *ProvisionAzureEventHubStep) retryOperationOnce(operation internal.ProvisioningOperation, wait time.Duration) (internal.ProvisioningOperation, time.Duration, error) {
+	return operation, wait, nil
+}
+
+// retryOperation retries an operation for at maxTime in retryInterval steps
+func (p *ProvisionAzureEventHubStep) retryOperation(operation internal.ProvisioningOperation, errorMessage string, retryInterval time.Duration, maxTime time.Duration, log logrus.FieldLogger) (internal.ProvisioningOperation, time.Duration, error) {
```
Comment (mszostok) [inline components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go]: @jasiu001 IMO those functions are quite generic and should be provided by the framework itself, WDYT?

Comment (jasiu001) [reply components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go]: I agree it could be used in other places/steps. Maybe the good place for this kind of method is `/internal/process/operation.go`

File: components/kyma-environment-broker/cmd/broker/main.go (hunk 21 above)
Comment (jasiu001) [reply components/kyma-environment-broker/cmd/broker/main.go]: in general, we do not support parallel steps yet but if you mean the same weight for step, you are right it could be the same weight (e.g. 2)

File: components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go (hunk 22 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go]: That sounds good! I think it is easier to read than just returning 3 arguments. I will place both methods to operation.go then.

File: components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+	log.Infof("Retrying for %s in %s steps", maxTime.String(), retryInterval.String())
+	if dur < maxTime {
+		return operation, retryInterval, nil
+	}
+	log.Errorf("Aborting after %s of failing retries", maxTime.String())
+	return p.operationManager.OperationFailed(operation, errorMessage)
+}
+
+func extractEndpoint(accessKeys eventhub.AccessKeys) string {
+	endpoint := strings.Split(*accessKeys.PrimaryConnectionString, ";")[0]
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go]: Good catch! The azure-sdk uses pointers everywhere 😂

File: components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go (hunk 22 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go]: We should have a follow up PR then to make use of these methods.

E.g. [here for RetryOperation](https://github.com/kyma-incubator/compass/blob/b13d9c2a0a4b6a2368f8cd7edef0112a92eb51f5/components/kyma-environment-broker/internal/process/provisioning/resolve_creds.go#L78
) and [here for retryOnce](https://github.com/kyma-incubator/compass/blob/b13d9c2a0a4b6a2368f8cd7edef0112a92eb51f5/components/kyma-environment-broker/internal/process/provisioning/resolve_creds.go#L102)

File: components/kyma-environment-broker/cmd/broker/main.go (hunk 21 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/cmd/broker/main.go]: So let's use the same weight, then we are future-proof if parallel steps feature is coming 😎

File: components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go (hunk 22 above)
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/process/provisioning/event_hub_step.go]: Done. I have the logger as method argument which does not look too nice. But since the logger comes the steps Run method and is not inside the steps structure for all steps which we have atm, it is the easiest solution for now.

File: components/kyma-environment-broker/internal/hyperscaler/azure/iam.go [hunk 24]
```diff
@@ -0,0 +1,47 @@
+package azure
+
+import (
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+)
+
+var (
+	armAuthorizer autorest.Authorizer
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/hyperscaler/azure/iam.go]: Totally agree.

[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/hyperscaler/azure/iam.go]: Done. Thanks for the comment. It was even confusing for me ;)

File: components/kyma-environment-broker/internal/hyperscaler/azure/iam.go [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+func GetResourceManagementAuthorizer(config *Config) (autorest.Authorizer, error) {
+	if armAuthorizer != nil {
+		return armAuthorizer, nil
+	}
+
+	environment, err := config.Environment()
+	if err != nil {
+		return nil, err
+	}
+	a, err := getAuthorizerForResource(config, environment.ResourceManagerEndpoint)
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/hyperscaler/azure/iam.go]: Done. Thanks for the comment. It was even confusing for me ;)

File: components/kyma-environment-broker/internal/hyperscaler/azure/group.go [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+)
+
+// getGroupsClient gets a client for handling of Azure ResourceGroups
+func getGroupsClient(config *Config) (resources.GroupsClient, error) {
+	client := resources.NewGroupsClient(config.subscriptionID)
+
+	authorizer, err := GetResourceManagementAuthorizer(config)
+	if err != nil {
+		return resources.GroupsClient{}, fmt.Errorf("failed to initialize authorizer with error: %v", err)
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/hyperscaler/azure/group.go]: Done

File: components/kyma-environment-broker/internal/hyperscaler/azure/iam.go [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+func GetResourceManagementAuthorizer(config *Config) (autorest.Authorizer, error) {
+	if armAuthorizer != nil {
+		return armAuthorizer, nil
+	}
+
+	environment, err := config.Environment()
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (nachtmaar) [reply components/kyma-environment-broker/internal/hyperscaler/azure/iam.go]: Done
